src/main/java/net/neoforged/neoforge/registries/datamaps/DataMapType.java

```diff
@@ -0,0 +1,167 @@
+import net.neoforged.neoforge.common.conditions.ConditionalOps;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A registry data map contains data-driven object that can be attached to a registry object. <p>
+ * Data maps are registered to the {@link RegisterDataMapTypesEvent}. <p>
+ * They are loaded from JSON files located at:
+ *
+ * <pre>
+ * <code>:mapNamespace/data_maps/:registryNamespace/:registryPath/:mapPath.json</code>
+ * </pre>
+ * 
+ * <p>
+ * The {@code registryNamespace} is omitted if it is {@value ResourceLocation#DEFAULT_NAMESPACE}. <br>
+ * The structure of the json file is as follows:
+ *
+ * <pre>
+ * <code>
+ *     {
+ *         "replace": false // If true, all previous data will be cleared.
+ *         // The values object is a map of registry entry ID / tag to data map values.
+ *         "values": {
+ *             "someobject:someid": {}, // The object being attached
+ *             "#somepath:sometag": {} // Tags are also supported. All objects in the tag will then receieve the same value
+ *         },
+ *         // Optional object. Entries specified here will be removed after the data of the current json file is attached
+ *         "remove": ["minecraft:carrot", "#minecraft:logs"]
+ *     }
+ * </pre>
+ *
+ * </code>
+ * Data maps support conditions both JSON-level and attachment-level through the {@value ConditionalOps#CONDITIONAL_VALUE_KEY} object.
+ * <p>
+ * Data maps may be synced by specifying a {@link #networkCodec()}. If the map is {@link #mandatorySync() mandatory},
+ * then vanilla clients (or any client that doesn't support this map) will not be able to connect.
+ *
+ * <p>
+ * Both datapack registries and normal, built-in registries support data maps.
+ *
+ * <p>
+ * You can access a data map using {@link net.neoforged.neoforge.registries.IRegistryExtension#getDataMap(DataMapType)} and {@link IWithData#getData(DataMapType)}. <br>
+ * You can usually go through {@linkplain net.minecraft.core.Holder#getData(DataMapType)} Holder} implementations in order to get the data of an object directly.
+ *
+ * @see AdvancedDataMapType for more functionality
+ */
+public sealed class DataMapType<T, R> permits AdvancedDataMapType {
+    private final ResourceKey<Registry<R>> registryKey;
+    private final ResourceLocation id;
+    private final Codec<T> codec;
+    private final @Nullable Codec<T> networkCodec;
+    private final boolean mandatorySync;
+
+    DataMapType(ResourceKey<Registry<R>> registryKey, ResourceLocation id, Codec<T> codec, @Nullable Codec<T> networkCodec, boolean mandatorySync) {
+        Preconditions.checkArgument(networkCodec != null || !mandatorySync, "Mandatory sync cannot be enabled when the attachment isn't synchronized");
+        Preconditions.checkArgument(registryKey != null, "registryKey must not be null");
+        Preconditions.checkArgument(codec != null, "codec must not be null");
+        Preconditions.checkArgument(id != null, "id must not be null");
```

Comment:
These null checks could just be done as setting the variable is done using `Objects#requireNonNull(Object, String)`